automerge-c/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+}
+
+/// \memberof AMdoc
+/// \brief Deallocates the storage for an `AMdoc` struct previously
+///        allocated by `AMcreate()` or `AMdup()`.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \pre \p doc must be a valid address.
+#[no_mangle]
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+pub unsafe extern "C" fn AMdestroy(doc: *mut AMdoc) {
+    if !doc.is_null() {
+        let doc: AMdoc = *Box::from_raw(doc);
+        drop(doc)
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Allocates storage for an `AMdoc` struct and initializes it by
+///        duplicating the `AMdoc` struct pointed to by \p doc.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \return A pointer to an `AMdoc` struct.
+/// \pre \p doc must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMdestroy()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+#[no_mangle]
+pub unsafe extern "C" fn AMdup(doc: *mut AMdoc) -> *mut AMdoc {
+    let doc = *Box::from_raw(doc);
+    let copy = doc.clone();
+    std::mem::forget(doc);
+    copy.into()
+}
+
+/// \memberof AMdoc
+/// \brief Sets a configuration property of an `AMdoc` struct.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] key A configuration property's key string.
+/// \param[in] value A configuration property's string value or `NULL`.
+/// \return A pointer to an `AMresult` struct containing no value.
+/// \pre \p doc must be a valid address.
+/// \pre \p key must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+/// key and value must be valid c strings
+#[no_mangle]
+pub unsafe extern "C" fn AMconfig(
+    doc: *mut AMdoc,
+    key: *const c_char,
+    value: *const c_char,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    let key = to_str!(key);
+    match key.as_str() {
+        "actor" => {
+            let actor = to_str!(value);
+            if let Ok(actor) = actor.try_into() {
+                doc.set_actor(actor);
+                AMresult::Ok.into()
+            } else {
+                AMresult::err(&format!("Invalid actor '{}'", to_str!(value))).into()
+            }
+        }
+        k => AMresult::err(&format!("Invalid config key '{}'", k)).into(),
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Get an `AMdoc` struct's actor ID value as a hexadecimal string.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \return A pointer to an `AMresult` struct containing a string value.
+/// \pre \p doc must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+#[no_mangle]
+pub unsafe extern "C" fn AMgetActor(_doc: *mut AMdoc) -> *mut AMresult {
+    unimplemented!()
```

Comment:
> Is this intended to be implemented soon...?

@alexjg, yes.

Reply:
Thinking more about this, I think we should still make a short note in the documentation that these are unimplemented. We should also note it on the README. I know this is all work in progress but it would be nice to be minimally confusing about the state of things for people who are taking a look at the `experiment`  branch and it becomes much more important once we move to `main`.

Reply:
> Is this intended to be implemented soon or should we document that it's unimplemented?

@alexjg, the `AMgetActor()` function is implemented in PR #310.

Reply:
Nice!